### ISSUE TITLE
[Request] Port Vuepress or Next.js to Hyperapp to show truly how it works

### ISSUE THREAD

**Issue by rajaraodv** [jorgebucaran/hyperapp#838]
All these counter, todo app etc are just examples. Port Vuepress or Next.js to Hyperapp to show truly how it works.

**Comment by frenzzy** [comment]
Or maybe it is better to create a clone of Facebook to show truly how it works?

**Comment by jorgebucaran** [comment]
@rajaraodv I have a bunch of [examples](https://github.com/jorgebucaran/hyperapp#examples) that I want to build in my todos. They will become available sooner or later. Just watch the repository.

Building a product similar to Vuepress sounds like a great idea. Thank you.

**Comment by rajaraodv** [comment]
Thanks, @jorgebucaran but none of those will really help people to get started in building a real-world app like Next.js, Nuxt, Sapper(Svelte) does.

**Comment by zaceno** [comment]
Perhaps true, but simply “porting Next.js to hyperapp” doesn’t exactly make sense either, since hyperapp and React have different architectural constraints. 

Sounds like you’re saying there should be a framework on top of hyperapp that solves many of the lower level problems for people. There may well be a case for something like that, but we won’t know, or know how to make something like that, until more real world apps are built on just hyperapp. 

If we tried now, we’d be just guessing at everything.

**Comment by SkaterDad** [comment]
I've been iterating on SPA framework over hyperapp v2 for a while now.   Currently not involving SSR, but it provides a nice structure and route-based code-splitting (`import()` ftw).

I have a decent sized app nearly public launch, and want to migrate it from hyperapp v1 to v2 using the framework.  Once I'm satisfied with how it works, I could probably open source it.

**Comment by jorgebucaran** [comment]
@rajaraodv I thought we were talking about building a large project like Vuepress to demonstrate how Hyperapp works. I  am skeptical that a large project is going to help anyone understand how to use Hyperapp, but I realize that some people prefer to learn by trial and error, others prefer to read the documentation or watch videos, and others prefer studying large, non-trivial projects as you suggest. I presume you may fall in this last category. A large project that "shows off" Hyperapp is a good idea in its own right, but is it going to help me understand how to use Hyperapp? I don't think so. I learned that there's a site https://hyperapp.rocks where people can submit projects they've built with Hyperapp. I don't know if the source code is available in all cases, but it's a start.

Now, I said I have several examples I want to work on in my todos like HNPWA and RealWorldApp didn't I? Thank you for your input, but I think projects of that size; not too large, but not too simple, are the perfect gateway to pick up a framework by doing if you want to put the theory aside. None of them are trivial, and any one of them should be sufficient to learn how Hyperapp works. Isn't what this issue is about?

I don't know what you mean by "show truly how it works". Other than what is documented, there's no "special" or "secret" way to use Hyperapp. Actually, it's quite simple, as in [Simplicity Matters by Rich Hickey](https://www.youtube.com/watch?v=rI8tNMsozo0) simple.

To summarize, do we want a Vuepress-like project built in Hyperapp? Yes, why not? Is one needed to teach how Hyperapp "truly" works? Most definitely not.

**Comment by jaeh** [comment]
i am building a static page / serverless app generator using hyperapp as the view layer.

### @magic
/core library and utilities (css-in-js, unit testing, cli wrapper, type checking and other tools i wrote for usage in @magic)

### @magic-modules
basically javascript objects that return state, actions and Components/Views.

also can include
* .style (javascript object with css styles, working on css string literal support for those that like it)
  [@magic/css](https://github.com/magic/css)
* .props (propType checking ala react),
* .server (faas lambdas now style [each lambda basically is a fn receiving request and response objects])
* .lib (client side library files to include in the bundle) (lib dependency resolution will be automated in the future)

### @magic-themes
currently only one theme there, the docs theme i use for all @magic docs

### @magic-libraries
tiny client side libraries that get included into your bundle if you need them

@magic is very opinionated, but (imho) allows an awesome workflow once used to it.

i am working on it full time (unless an urgent job comes up),
serverless/faas should be usable within the next 1-2 months,
using now.sh (almost done), self-hosted services (later) or our @magic cloud (even later) ;)

some pages done with it:
#### https://magic.github.io/core (main docs)
https://magic.github.io (list of all modules, themes and libraries)
https://jaeh.at (custom url, repo is https://github.com/jaeh/jaeh.at)
https://wizardsatwork.at 
please make sure to look at the filesizes of the loaded assets when looking at the pages ;)

my plan is to release 0.0.1 on npm within the next 1-2 weeks,
then continue happily hacking forward and release 0.1.0 sometime this summer,
at which point the apis will have settled and there should be a cms like admin interface.

i would love to get feedback :)

**Comment by jorgebucaran** [comment]
@jaeh Would be nice to see it ported to V2 since V1 obsolete now.

**Comment by jaeh** [comment]
thats on my list to try to get done before 0.1.0,
last time i tried a few months ago i could not get server side rendering to work with v2,
do you have a link to a working example repo with hyperapp-render?

**Comment by jorgebucaran** [comment]
@jaeh Sorry, I don't have any (hyperapp-render is not a core package—technically there's no official server-side rendering for Hyperapp).

**Comment by SkaterDad** [comment]
@jaeh `hyperapp-render` already supports V2 according to the docs.

```js
import { renderToString, /* or */ renderToStream  } from 'hyperapp-render'
import { view, intialState } from './myapp.js'

const html = renderToString(view(initialState))
```

**Comment by jaeh** [comment]
nice, sorry for even asking, should have just tried what the docs say.

i generate the client side bundle and wrap the module state and actions, so i have pushed upgrading hyperapp to the point where i have no other todos for 0.1. 
this coincides with the upgrade to es6 modules for the client, which will allow exposing both mjs and js files for the index.js file.

i also expect actions and state to be easier to handle with v2, so i really look forward to using it :)

**Comment by rajaraodv** [comment]
@zaceno @jorgebucaran The subject is incorrect. I meant create a project like Next.js to show the power of Hyperapp. 

Currently I've built two production apps https://testautomationu.com  (with 10s of 1000s of users) and a docs app: https://applitools.com/tutorials both built on top of Vuepress. I chose Vuepress not only because Vuejs is relatively simple, but Vuepress had everything (like markdown compilation, routing, SSR, hot reloading etc) provided out-of-the-box and helped me rapidly build things and focus on getting users. 

My feeling is that just hyperapp without projects like vuepress or Next.js that's built on top of Hyperapp, will not be that useful for a lot of projects where time is limited. So people like me who'd like to use things like Hyperapp will choose something that already provides all those things and move on.

**Comment by jorgebucaran** [comment]
@rajaraodv As I said, building a product similar to Vuepress sounds like a great idea. Why not hire someone to do it? You may also support me on Patreon if you're so inclined. As an issue request, this is a tad random, but my issues are open to all kind of requests, so thank you for bringing it up anyway.

I'm also working on a Starter Kit that I plan to use with my projects. It may not be as sophisticated as `create-react-app` but will work for me:


> _**Everything you need to get started building applications with Hyperapp.**_
>
> _The Hyperapp Starter Kit is your gateway to higher-order levels of the imagination. This box contains the essential concepts to set you off on your journey through the perilous land of functional frontend programming plus everything you need to create ambitious, next-generation JavaScript applications without side effects. Your hair looks awesome, by the way. Let's rock!_
</i>

**Comment by jorgebucaran** [comment]
Seriously, @Magic sounds awesome. Wish you good luck with your project @jaeh.

**Comment by jaeh** [comment]
this is a bit off topic, but since mentioned magic here i thought i should also post this note:
tldr:
i am currently rewriting all of my libs to use es modules, unfortunately, this breaks magic (and esp. the docs) until all libraries are translated.


@jorgebucaran's comment regarding v2 got me going trying to get esmodule coverage working with my unit testing library.
which, after about 2 years of having the mjs branch ready, i found a way to do.

i am already at the point where @magic/core#mjs works locally, just wanted to mention that the magic ecosystem will be wonky and documentation is lying until i comment here again, at which point hyperapp v2 should be landing in magic as well :)

thanks for nudging me re v2 (and the nice feedback) :)
